allocation.md

```diff
@@ -1,53 +1,406 @@
-  ramEfficiency: .56                     // percentage of RAM requested that is utilized, weighted by cost
-  sharedCost: 0
-  start: "2020-09-13T00:00:00+01:00"     // beginning of window
-  totalCost: 0.024                       // sum of all costs
-  totalEfficiency: .72                   // percentage of RAM + CPU equested that is utilized, weighted by cost
+# Allocation
+
+The Allocation API is the preferred way to query for costs and resources allocated to Kubernetes workloads, and optionally aggregated by Kubernetes concepts like `namespace`, `controller`, and `label`. Data is served from one of [Kubecost's ETL pipelines](https://github.com/kubecost/docs/blob/master/allocation-api.md#caching-overview). The endpoint is available at the URL:
+```
+http://<kubecost>/model/allocation
+```
+
+> **NOTE**
+>
+> Throughout, we use `localhost:9090` as the default Kubecost URL, but your Kubecost instance may be exposed by a service or ingress. To reach Kubecost at port 9090, run: `kubectl port-forward deployment/kubecost-cost-analyzer -n kubecost 9090`
+
+## Quick start
+
+Request allocation data for each 24-hour period in the last three days, aggregated by namespace:
+```
+$ curl http://localhost:9090/model/allocation \
+  -d window=3d \
+  -d aggregate=namespace \
+  -d accumulate=false \
+  -d shareIdle=false
+  -G
+```
+```json
+{
+  "code": 200,
+  "data": [
+    {
+      "__idle__": { ... },
+      "default": { ... },
+      "kube-system": { ... },
+      "kubecost": { ... }
+    },
+    {
+      "__idle__": { ... },
+      "default": { ... },
+      "kube-system": { ... },
+      "kubecost": { ... }
+    },
+    {
+      "__idle__": { ... },
+      "default": { ... },
+      "kube-system": { ... },
+      "kubecost": { ... }
+    },
+    {
+      "__idle__": { ... },
+      "default": { ... },
+      "kube-system": { ... },
+      "kubecost": { ... }
+    }
+  ]
+}
+```
+
+Note: querying for "3d" will likely return a range of four sets because the queried range will overlap with four precomputed 24-hour sets, each aligned to the configured timezone. For instance, querying "3d" on 2021/01/04T12:00:00 will return:
+- 2021/01/04 00:00:00 until 2021/01/04T12:00:00 (now)
+- 2021/01/03 00:00:00 until 2021/01/04 00:00:00
+- 2021/01/02 00:00:00 until 2021/01/03 00:00:00
+- 2021/01/01 00:00:00 until 2021/01/02 00:00:00
+
+See [Querying](#querying) for the full list of arguments and [Examples](#examples) for more example queries.
+
+## Allocation schema (version 1.76)
+Field | Description
+---: | :---
+name | Name of each relevant Kubernetes concept described by the allocation, delimited by slashes, e.g. "cluster/node/namespace/pod/container"
+properties | Map of name-to-value for all relevant property fields, including: `cluster`, `node`, `namespace`, `controller`, `controllerKind`, `pod`, `container`, `labels`, `annotation`, etc.
+window | Period of time over which the allocation is defined.
```

Comment:
If this is the same as the window of the precomputed 24 hr sets mentioned earlier, mention that in this description because there's also the query parameter `window` with different meaning.

Reply:
Actually, this is probably my own repeated confusion about `window`. I assume this is computed based on the `window=3d` query parameter, and then `start` and `end` are what I was talking about in the first comment. Could we then modify this definition to be more clear about what "over which the allocation is defined" means?

Something like: "Period of time (format: `FORMAT`) defined by the time of query and the `window` parameter. This will be the same for every object in `data[]`."

Reply:
"window" is one of the most loaded terms we have, so unfortunately it does get confusing. I like where you're headed with that suggestion, but even that isn't quite right because for "3d" you will get three different sets, the windows of which will sum to the full queried window, but it won't be the same for every object in `data[]`.

Reply:
Oh, shoot. So `window` is actually the duration "precomputed 24 hr set" mentioned earlier, and then `start` and `end` are the actual bookends of the **allocation**, yeah?

Reply:
Keeping this all straight is a bit tricky still :sweat_smile:

Reply:
How does this sound:
> Period of time (format: `"start": "2009-01-02T15:04:05Z07:00", "end": "2009-01-02T15:04:05Z07:00"`) within the queried window, over which the allocation is defined. Will be the same for each allocation in any given object within the `data` array.

Reply:
Yeah, I think that sounds good. :+1: